File: tests/cpu/seats/roster-rows.test.ts
import { describe, it, expect } from 'vitest';
import { makeMap, makeState, tileAtCoords, settleAt, grantTechs } from '../helpers';
import { emptySeat } from '../../../cpu/core/seats';
import { seatPhase } from '../../../cpu/core/phase';
import { computeCityStats } from '../../../cpu/core/city';
import { prodMultFor, getModifiers } from '../../../cpu/core/effects';
import { routeYieldsInternational, tradeCapacity, rosterRouteCapacity } from '../../../cpu/core/trade';
import { neighbors } from '../../../world/hex';
import { CIV_LEADERS } from '../../../cpu/data/seats';
import { PROD_MULT_ROWS, DISTRICT_ADJ_ROWS, INTL_ROUTE_YIELD_ROWS, ROUTE_CAPACITY_ROWS } from '../../../cpu/data/civilizations';
import type { City, GameState, Tile } from '../../../cpu/core/types';

/**
 * THE ROSTER'S DATA ROWS (CIV6, the install's TraitModifiers): production
 * percentages, Meiji Restoration's district adjacency, Radio Oranje's route
 * culture, and the capacity rows of Nîhithaw and Founder of Carthage — one
 * clause per assertion, on the site that pays it.
 */
const seatRow = (civ: string) => CIV_LEADERS.findIndex((l) => l.civ === civ);

function placeDistrict(state: GameState, city: City, type: 'HOLY_SITE' | 'CAMPUS' | 'GOVERNMENT_PLAZA', tile: Tile): void {
  tile.district = type;
  tile.districtComplete = true;
  city.districts.push({ type, tileIndex: tile.index } as City['districts'][number]);
}

describe('the production rows', () => {
  it('prodMultFor multiplies the rows that name the item', () => {
    expect(prodMultFor(PROD_MULT_ROWS.filter((r) => r.civ === 'ENGLAND'), { building: 'WORKSHOP', district: 'INDUSTRIAL_ZONE' })).toBe(1.2);
    expect(prodMultFor(PROD_MULT_ROWS.filter((r) => r.civ === 'ENGLAND'), { building: 'MONUMENT', district: 'CITY_CENTER' })).toBe(1);
    expect(prodMultFor(PROD_MULT_ROWS.filter((r) => r.civ === 'GEORGIA'), { building: 'ANCIENT_WALLS', district: 'CITY_CENTER' })).toBe(1.5);
    expect(prodMultFor(PROD_MULT_ROWS.filter((r) => r.civ === 'OTTOMAN'), { promoClass: 'SIEGE' })).toBe(1.5);
    expect(prodMultFor(PROD_MULT_ROWS.filter((r) => r.civ === 'OTTOMAN'), { promoClass: 'MELEE' })).toBe(1);
  });

  it('pay on the queue: England +20% on a Workshop, Georgia +50% on walls, the Ottomans +50% on a Catapult', () => {
    const run = (civ: string, item: City['queue'][number]): number => {
      const state = makeState(makeMap(12, 12, 'GRASSLAND'));
      state.seats[0].civ = seatRow(civ);
      const city = settleAt(state, tileAtCoords(state.map, 6, 6).index, 0);
      city.queue.push({ ...item }); // a fresh entry per run: the phase writes its progress
      seatPhase(state);
      return city.queue[0].progress;
    };
    const ws = { kind: 'building', building: 'WORKSHOP', progress: 0 } as City['queue'][number];
    expect(run('ENGLAND', ws)).toBe(run('AMERICA', ws) * 1.2); // America holds no production row
    const walls = { kind: 'building', building: 'ANCIENT_WALLS', progress: 0 } as City['queue'][number];
    expect(run('GEORGIA', walls)).toBe(run('AMERICA', walls) * 1.5);
    const cat = { kind: 'unit', unit: 'CATAPULT', progress: 0 } as City['queue'][number];
    expect(run('OTTOMAN', cat)).toBe(run('AMERICA', cat) * 1.5);
    expect(run('OTTOMAN', ws)).toBe(run('AMERICA', ws));
  });
});

describe('Meiji Restoration', () => {
  it('adds +1 to a district per adjacent district, Japan alone', () => {
    expect(DISTRICT_ADJ_ROWS.length).toBe(6);
    const faithOf = (civ: string): number => {
      const state = makeState(makeMap(12, 12, 'GRASSLAND'));
      state.seats[0].civ = seatRow(civ);
      const city = settleAt(state, tileAtCoords(state.map, 6, 6).index, 0);
      const around = neighbors(state.map, state.map.tiles[city.centerIndex]);
      const hs = around[0];
      placeDistrict(state, city, 'HOLY_SITE', hs);
      const beside = neighbors(state.map, hs).find((t) => t.index !== city.centerIndex && !t.district)!;
      placeDistrict(state, city, 'CAMPUS', beside);
      expect(getModifiers(state, 0).districtAdjacencyAdd.HOLY_SITE?.length ?? 0).toBe(civ === 'JAPAN' ? 1 : 0);
      // the centre is an adjacent district too: +1 per neighbour holding one
      const n = neighbors(state.map, hs).filter((t) => t.district === 'CITY_CENTER' || (t.district && t.districtComplete)).length;
      return computeCityStats(state, city).breakdown.districts.faith - (civ === 'JAPAN' ? n : 0);
    };
    expect(faithOf('JAPAN')).toBe(faithOf('AMERICA'));
  });
});

describe('Radio Oranje', () => {
  it('pays +2 Culture on an international route the Netherlands send', () => {
    expect(INTL_ROUTE_YIELD_ROWS.length).toBe(1);
    const state = makeState(makeMap(12, 12, 'GRASSLAND'));
    state.seats.push(emptySeat(1));
    state.seats[0].civ = seatRow('NETHERLANDS');
    state.seats[1].civ = seatRow('ROME');
    settleAt(state, tileAtCoords(state.map, 3, 3).index, 0);
    const dest = settleAt(state, tileAtCoords(state.map, 9, 9).index, 1);
    expect(getModifiers(state, 0).leader).toBe('WILHELMINA');
    expect(routeYieldsInternational(state, dest, 0).culture).toBe(2);
    expect(routeYieldsInternational(state, dest, 1).culture).toBe(0);
  });
});

describe('the capacity rows', () => {
  it("Nîhithaw: +1 at Pottery once a capital stands; Founder of Carthage: +1 per plaza and tier", () => {
    expect(ROUTE_CAPACITY_ROWS.length).toBe(5);
    const cree = makeState(makeMap(12, 12, 'GRASSLAND'));
    cree.seats[0].civ = seatRow('CREE');
    expect(rosterRouteCapacity(cree, 0)).toBe(0); // no capital yet
    settleAt(cree, tileAtCoords(cree.map, 6, 6).index, 0);
    expect(rosterRouteCapacity(cree, 0)).toBe(0); // no Pottery yet
    grantTechs(cree, 'POTTERY');
    expect(rosterRouteCapacity(cree, 0)).toBe(1);
    expect(tradeCapacity(cree, 0)).toBe(1);
    const dido = makeState(makeMap(12, 12, 'GRASSLAND'));
    dido.seats[0].civ = seatRow('PHOENICIA');
    expect(getModifiers(dido, 0).leader).toBe('DIDO');
    const city = settleAt(dido, tileAtCoords(dido.map, 6, 6).index, 0);
    expect(rosterRouteCapacity(dido, 0)).toBe(0);
    placeDistrict(dido, city, 'GOVERNMENT_PLAZA', neighbors(dido.map, dido.map.tiles[city.centerIndex])[0]);
    expect(rosterRouteCapacity(dido, 0)).toBe(1);
    city.buildings.push('ANCESTRAL_HALL');
    expect(rosterRouteCapacity(dido, 0)).toBe(2);
    city.buildings.push('GRAND_MASTERS_CHAPEL');
    expect(rosterRouteCapacity(dido, 0)).toBe(3);
    const rome = makeState(makeMap(12, 12, 'GRASSLAND'));
    rome.seats[0].civ = seatRow('ROME');
    settleAt(rome, tileAtCoords(rome.map, 6, 6).index, 0);
    grantTechs(rome, 'POTTERY');
    expect(rosterRouteCapacity(rome, 0)).toBe(0);
  });
});

File: tests/gpu/roster_rows_test.py
"""THE ROSTER'S DATA ROWS — the GPU half.

    npm run seed && npm run export        # (once) writes seeder/worlds/
    python tests/gpu/roster_rows_test.py

The TS twin is tests/cpu/seats/roster-rows.test.ts.

CIV6 (the install's TraitModifiers): the production percentages (England,
Georgia, the Netherlands, the Ottomans), Meiji Restoration's district
adjacency, Radio Oranje's route culture, and the capacity rows of Nîhithaw
and Founder of Carthage — each on the site that pays it, the seat's
civilization or leader alone.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

import torch

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent / "gpu"))
sys.path.insert(0, str(Path(__file__).resolve().parent))
from core import BatchSim, load_rules, load_fixture, fixture_paths
from warmup import settle_all

B0 = 0
RULES = json.loads((Path(__file__).resolve().parent.parent.parent
                    / "seeder" / "worlds" / "rules.json").read_text())
UNITS = [u["id"] for u in RULES["units"]]
BLDGS = [b["id"] for b in RULES["buildings"]]
TECHS = [t["id"] for t in RULES["techs"]]


def play(sim, row: int, name):
    if name is None:
        sim.row_civ[0, row] = -1
        sim.row_leader[0, row] = -1
    else:
        ci = sim._civ_ids.index(name)
        sim.row_civ[0, row] = ci
        sim.row_leader[0, row] = sim._pair_civ.index(ci)
    sim._eff_version += 1
    sim._gen_ver += 1
    sim._bldg_version += 1


def fresh(rules, path) -> BatchSim:
    sim = BatchSim([load_fixture(path)], rules, device="cpu", dtype=torch.float64)
    for r, name in enumerate(("ROME", "EGYPT", "NORWAY")):
        play(sim, r, name)
    return settle_all(sim)


def district_idx(sim, name: str) -> int:
    return next(i for i, d in enumerate(sim.districts_cat) if d["id"] == name)


def produce(sim, row: int, code: int) -> float:
    sim.city_current[B0, row, 0, 0] = code
    sim.city_qtile[B0, row, 0, 0] = -1
    sim.city_progress[B0, row, 0, 0] = 0
    sim.city_cost[B0, row, 0, 0] = 100000
    sim._eff_version += 1
    sim._seat_city_produce(row, torch.tensor([0]), torch.tensor([True]), torch.tensor([20.0], dtype=torch.float64))
    return float(sim.city_progress[B0, row, 0, 0])


# ---------------------------------------------------------------------------


def test_wire(rules, path) -> None:
    sim = fresh(rules, path)
    assert len(sim._prod_mult_rows) == 6 and len(sim._district_adj_rows) == 6
    assert len(sim._intl_route_rows) == 1 and len(sim._route_cap_rows) == 5
    assert all(r[4] >= 0 for r in sim._prod_mult_rows if r[2] < 0 and r[3] < 0), "a unit-class row lost its class"
    print("  1 wire OK — 6 + 6 + 1 + 5 rows")


def test_production(rules, path) -> None:
    ws, walls, cat = BLDGS.index("WORKSHOP"), BLDGS.index("ANCIENT_WALLS"), UNITS.index("CATAPULT")

    def run(as_civ: str, code: int) -> float:
        sim = fresh(rules, path)
        play(sim, 0, as_civ)
        return produce(sim, 0, code)

    base_ws = run("AMERICA", ws)  # America holds no production row
    assert base_ws > 0 and abs(run("ENGLAND", ws) - base_ws * 1.2) < 1e-9, "Workshop of the World"
    assert abs(run("OTTOMAN", ws) - base_ws) < 1e-9, "the Ottomans paid a Workshop"
    base_w = run("AMERICA", walls)
    assert abs(run("GEORGIA", walls) - base_w * 1.5) < 1e-9, "Strength in Unity"
    base_c = run("AMERICA", sim_code := sim_unit_code(fresh(rules, path), cat))
    assert abs(run("OTTOMAN", sim_code) - base_c * 1.5) < 1e-9, "Great Turkish Bombard"
    print("  2 production rows OK — England x1.2, Georgia x1.5, the Ottomans x1.5")


def sim_unit_code(sim, utype: int) -> int:
    return sim.UNIT_BASE + utype


def test_meiji(rules, path) -> None:
    def faith_at(as_civ: str) -> float:
        sim = fresh(rules, path)
        play(sim, 0, as_civ)
        hs, campus = district_idx(sim, "HOLY_SITE"), district_idx(sim, "CAMPUS")
        ctr = int(sim.city_center[B0, 0, 0])
        ring = [int(x) for x in sim.neigh[ctr].tolist() if x >= 0 and not bool(sim.water[B0, x]) and bool(sim.passable[B0, x])]
        site = ring[0]
        beside = next(x for x in sim.neigh[site].tolist() if x >= 0 and x != ctr and not bool(sim.water[B0, x]) and bool(sim.passable[B0, x]))
        for t, di in ((site, hs), (beside, campus)):
            sim.tile_seat[B0, t] = 0
            sim.district[B0, t] = di
            sim.district_complete[B0, t] = True
            sim.city_dist_tile[B0, 0, 0, di] = t
        sim._tile_owner_ver += 1
        sim._eff_version += 1
        # the centre is an adjacent district too: +1 per neighbour holding one
        n = sum(1 for x in sim.neigh[site].tolist() if x >= 0 and (int(sim.centre_slot_at[B0, x]) >= 0
                or (int(sim.district[B0, x]) >= 0 and bool(sim.district_complete[B0, x]))))
        return float(sim._district_adj_floor(hs)[B0, site]) - (n if as_civ == "JAPAN" else 0)

    assert faith_at("JAPAN") == faith_at("AMERICA"), "Meiji Restoration"
    print("  3 Meiji Restoration OK — +1 per adjacent district")


def test_radio_oranje(rules, path) -> None:
    sim = fresh(rules, path)
    play(sim, 0, "NETHERLANDS")
    assert bool(sim._row_leads(0, "WILHELMINA")[B0])
    sim.seat_routes[B0, 0, 0, 0] = int(sim.city_id[B0, 0, 0])
    sim.seat_routes[B0, 0, 0, 1] = -1
    sim.seat_route_dseat[B0, 0, 0] = 1
    sim.seat_route_dcity[B0, 0, 0] = int(sim.city_id[B0, 1, 0])
    sim.seat_route_exp[B0, 0, 0] = int(sim.turn) + 20
    sim._eff_version += 1
    inc = sim._seat_route_income(0)
    assert inc is not None and abs(float(inc[B0, 0, 4]) - 2.0) < 1e-9, float(inc[B0, 0, 4])
    play(sim, 0, "ROME")
    assert float(sim._seat_route_income(0)[B0, 0, 4]) == 0.0, "the culture outlived Wilhelmina"
    print("  4 Radio Oranje OK — +2 Culture on the international leg")


def test_capacity(rules, path) -> None:
    sim = fresh(rules, path)
    play(sim, 0, "CREE")
    pottery = TECHS.index("POTTERY")
    sim.civ_techs[B0, 0, pottery] = False
    assert int(sim._roster_route_capacity(0)[B0]) == 0
    sim.civ_techs[B0, 0, pottery] = True
    assert int(sim._roster_route_capacity(0)[B0]) == 1, "Nîhithaw"
    before = int(sim._trade_capacity(0)[B0])
    play(sim, 0, "ROME")
    assert int(sim._trade_capacity(0)[B0]) == before - 1, "the capacity outlived the Cree"
    sim = fresh(rules, path)
    play(sim, 0, "PHOENICIA")
    assert bool(sim._row_leads(0, "DIDO")[B0])
    assert int(sim._roster_route_capacity(0)[B0]) == 0
    plaza = district_idx(sim, "GOVERNMENT_PLAZA")
    ctr = int(sim.city_center[B0, 0, 0])
    t = next(int(x) for x in sim.neigh[ctr].tolist() if x >= 0 and not bool(sim.water[B0, x]) and bool(sim.passable[B0, x]))
    sim.district[B0, t] = plaza
    sim.district_complete[B0, t] = True
    sim.city_dist_tile[B0, 0, 0, plaza] = t
    sim._eff_version += 1
    assert int(sim._roster_route_capacity(0)[B0]) == 1, "the plaza"
    sim.city_bldg[B0, 0, 0, BLDGS.index("ANCESTRAL_HALL")] = True
    assert int(sim._roster_route_capacity(0)[B0]) == 2, "tier 1"
    sim.city_bldg[B0, 0, 0, BLDGS.index("GRAND_MASTERS_CHAPEL")] = True
    assert int(sim._roster_route_capacity(0)[B0]) == 3, "tier 2"
    print("  5 capacity rows OK — Pottery with a capital; the plaza and its tiers")


def main() -> int:
    rules = load_rules()
    path = fixture_paths()[0]
    test_wire(rules, path)
    test_production(rules, path)
    test_meiji(rules, path)
    test_radio_oranje(rules, path)
    test_capacity(rules, path)
    print("BATTERY OK roster_rows")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
